### PR TITLE
Implement the PATCH method

### LINKED ISSUE #1941: Add support for the PATCH method
Right now the LXD API only supports GET, POST, PUT and DELETE. It would be very beneficial to support PATCH for all endpoints which support PUT today.

### PR DESCRIPTION
**Pull request by stgraber**
Closes #1941

Signed-off-by: Stéphane Graber <stgraber@ubuntu.com>

### CODE REVIEW

File: lxd/container_patch.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+	}
+
+	// Validate the ETag
+	etag := []interface{}{c.Architecture(), c.LocalConfig(), c.LocalDevices(), c.IsEphemeral(), c.Profiles()}
+	err = etagCheck(r, etag)
+	if err != nil {
+		return PreconditionFailed(err)
+	}
+
+	body, _ := ioutil.ReadAll(r.Body)
```
Comment (tych0) [inline lxd/container_patch.go]: Can't we get a network error here?

File: lxd/container_patch.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+		architecture = c.Architecture()
+	} else {
+		architecture, err = shared.ArchitectureId(req.Architecture)
+		if err != nil {
+			architecture = 0
+		}
+	}
+
+	// Check if ephemeral was passed
+	_, err = reqRaw.GetBool("ephemeral")
```
Comment (tych0) [inline lxd/container_patch.go]: Should we error out here, if we don't actually allow setting this property?

Comment (tych0) [inline lxd/container_patch.go]: Oh, no, derp. We do, never mind :)

File: lxd/container_patch.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	if err := json.NewDecoder(rdr2).Decode(&req); err != nil {
+		return BadRequest(err)
+	}
+
+	if req.Restore != "" {
+		return BadRequest(fmt.Errorf("Can't call PATCH in restore mode."))
+	}
+
+	// Check if architecture was passed
+	var architecture int
```
Comment (sean-jc) [inline lxd/container_patch.go]: Can someone explain (or point me at documentation) how the architecture of a container can be updated without everything exploding?  I'm trying, and failing, to wrap my head around this concept.

**Pull request merged**